server: init functional tests

phymbert opened the pull request:
## Motivation
Tests has been listed in #4216 as an improvements request.

The idea is to ensure all server routes are working properly using the [Gherkin](https://cucumber.io/docs/gherkin/reference/) language to define test cases following [BDD](https://en.wikipedia.org/wiki/Behavior-driven_development) approach. It is designed to be human-readable, and describes use cases relating to a software system.

### Example
```feature
  @llama.cpp
  Scenario: Multi users
    Given a prompt:
      """
      Write a very long story about AI.
      """
    And a prompt:
      """
      Write another very long music lyrics.
      """
    And 32 max tokens to predict
    Given concurrent completion requests
    Then the server is busy
    And  all slots are busy
    Then the server is idle
    And  all slots are idle
    Then all prompts are predicted
```

## Proposed changes

A CI workflow is triggered which builds and starts the server in background, then test scenario are launched with python.
A very [small model](https://huggingface.co/ggml-org/models/blob/main/tinyllamas/stories260K.gguf) is used to quickly generate responses and a fixed `seed` is set to ensure reproducibility.

The Gherkin glue is written in python using [behave](https://behave.readthedocs.io/en/latest/).

## Restriction
This is not designed to assess performance of the server.

## Expected scenario:

(from @ngxson [comment](https://github.com/ggerganov/llama.cpp/issues/4216#issuecomment-1905758085))

- [x] health and slots endpoints
- [x] completion endpoint
- [x] OAI compatible chat completion requests w/ and without streaming
- [ ] multimodal prompts
- [x] passing multi users scenario
- [x] multi users scenario on OAI compatible endpoint with streaming
- [x] multi users with total number of tokens to predict exceeds the KV Cache size
- [x] server wrong usage scenario, like in #3969 
- [x] slots shifting
- [x] continuous batching
- [x] embeddings endpoint
- [ ] embeddings endpoint with image
- [x] multi user embedding endpoint: #5655
- [x] OpenAI-compatible embeddings API
- [x] tokenize endpoint
- [ ] infill endpoint
- [x] CORS and api key scenario @Azeirah
- [ ] Upon receiving an non-completed unicode, the json converter crashes the whole process. @ngxson 

Example of passing GitHub workflow can be found [here](https://github.com/ggerganov/llama.cpp/actions/runs/8029939149).

## TODO:
- [x] Use different build cmake matrix of the server to run tests:         sanitizer: [ADDRESS, THREAD, UNDEFINED], build_type: [Debug, Release],   build: [noavx, avx2, avx, avx512, clblast, openblas, kompute, vulkan, cublas]. Even if it works on docker ubuntu with CPU, it is interesting to see if it still works when no GPU device are available.
- [x] start the server within the scenario background instead of starting it manually to pass custom parameters
- [x] make the server binary path configurable
- [x] fix `/slots` and `/health` endpoint to properly access slots data over the `queue_tasks`: #5634
- [x] [fix](https://github.com/phymbert/llama.cpp/blob/test/server-add-ci-test/examples/server/server.cpp#L1427) slots[].state in health endpoint may be incoherent with the total slots idle under race condition 
- [x] change the ci build trigger
- [x] use asyncio and aiohttp to trigger concurrent http request
- [x] [fix](https://github.com/phymbert/llama.cpp/blob/test/server-add-ci-test/examples/server/tests/features/parallel.feature#L54) async openai http client with streaming.
- [ ] sometimes health check requests are triggered after completions request, making the scenario unstable
- [ ] capture stdout / stderr of server_process maybe saved to a second file called llama-stdout.log @ngxson

phymbert committed:
server: init functional test

ggerganov commented:
I'm quite out of depth here, but if you can figure out a way to add `server` tests it would be awesome. I've sent you a collaborator invite

phymbert commented:
@RafaAguilar @ngxson As you were part of the tests related discussion, do you feel OK with the proposed approach here ? If so, I will continue with asynchronous request and multi users scenario

ngxson commented:
Great idea, thanks for starting this PR. Some suggestions:
1. Since the number of test cases is not very big, can we reduce number of files? (so that future contributors can find things more easily)
2. Would be nice if we have a .sh script that runs `server -m ...` and `python -m behave` all at once. It's again to be easier for future contributors. It can also be useful when we want to dockerize the test script in the future.
3. Note to myself: maybe I can fine tune the bloom 560m model for using in this test. The smallest usable gguf that we can find on hf is currently `tinyllama-2-1b-miniguanaco.Q2_K.gguf`

Azeirah commented:
Excellent! I'll be reading this PR today and see if I can add a test or help out in some way :)

Azeirah commented:
> Great idea, thanks for starting this PR. Some suggestions:
> 
> 1. Since the number of test cases is not very big, can we reduce number of files? (so that future contributors can find things more easily)
> 2. Would be nice if we have a .sh script that runs `server -m ...` and `python -m behave` all at once. It's again to be easier for future contributors. It can also be useful when we want to dockerize the test script in the future.
> 3. ~Note to myself: maybe I can fine tune the bloom 560m model for using in this test. The smallest usable gguf that we can find on hf is currently `tinyllama-2-1b-miniguanaco.Q2_K.gguf`~ Tried finetuning `ahxt/LiteLlama-460M-1T` but the result is unusable.

Would it be possible to train a nonsensical 1m param model potentially? It should be really cheap and fast even on commodity hardware. These tests aren't meant to interact in any meaningful way with the output anyway.

ngxson commented:
@Azeirah Yes it's possible, but the problem is that these models never want to output EOS token (to terminate the output) . It's also possible to rely on the `n_predict` to stop the generation after X tokens.
Another problem is that small model tends to output invalid bytes instead of words. Maybe I need to limit the usable tokens in its vocab.
Anyway, I'll look into this this week, it's still a good exercise for me to train a model from zero.

Azeirah commented:
> @Azeirah Yes it's possible, but the problem is that these models never want to output EOS token (to terminate the output) . It's also possible to rely on the `n_predict` to stop the generation after X tokens.
> Another problem is that small model tends to output invalid bytes instead of words (because part of llama vocab is bytes, which allow it to do unicode). Maybe I need to limit the usable tokens in its vocab.
> Anyway, I'll look into this this week, it's still a good exercise for me to train a model from zero.

Fair enough, I think it should be doable to make a model that behaves well enough. Potentially it could be trained explicitly to bias EOS whaha. I agree it would be a fun exercise, unfortunately I have a 7900xtx and I believe it cannot be used to train :(

In addition to that, we of course have no clue what kind of hardware these tests will be ran on, but if it's a virtual core on a xeon or some or other maybe we can try compiling openBLAS? I'm not sure if it'd be even worth investigating depending on the speedup and the variety of weird hardware you could get on Github actions. No clue what kind of control over the underlying (virtualised) hardware you'd get there.

----

Other than that, I think it's fine that the tests are in separate files. It's kinda just how behave is meant to be used, each feature is one file. Different related scenarios belong to one feature.

ngxson commented:
Also one case that I have never tested before is invalid unicode.

In my personal project (which uses llama.h), on receiving responses via `llama_token_to_piece`, I pass it to `nlohmann/json` to convert it to json string. That's the same thing we're using in server example. Upon receiving an non-completed unicode, the json converter crashes the whole process.

Would be nice if someone can test if it's the case for server.cpp (which `stream=True` for example)

ngxson commented:
@Azeirah I believe the hosted runner of github is Xeon with shared CPU cores. The performance is not meant to be consistent though. I believe that it cannot use anything better than AVX2.

For training, I'm using a GTX 1660 Ti. I initially purchased it for gaming 2 years ago, but who knows that now I need more VRAM than that :'( Back then, the dealer proposed me a 3080 Ti with a fairly good price, but I refused. Nowadays, for anything bigger than 1B, I need to rent a VPS on google cloud, it's more or less the same price with colab notebooks, but more flexible and have persistent storage.

phymbert committed:
server: tests: reduce number of files, all in one tests shell script

phymbert commented:
> Great idea, thanks for starting this PR. Some suggestions:
> 
> 1. Since the number of test cases is not very big, can we reduce number of files? (so that future contributors can find things more easily)
> 2. Would be nice if we have a .sh script that runs `server -m ...` and `python -m behave` all at once. It's again to be easier for future contributors. It can also be useful when we want to dockerize the test script in the future.
> 3. ~Note to myself: maybe I can fine tune the bloom 560m model for using in this test. The smallest usable gguf that we can find on hf is currently `tinyllama-2-1b-miniguanaco.Q2_K.gguf`~ Tried finetuning `ahxt/LiteLlama-460M-1T` but the result is unusable.

Done :+1:

phymbert commented:
@ggerganov @ngxson Any idea on how to improve the prompt eval time on the github runners ? Should we give a try to OpenBLAS ?

phymbert committed:
server: tests: fix ci workflow

ngxson commented:
@phymbert Can you try this model instead? (pay attention to set `n_predict` or `max_tokens`, because the model never outputs EOS token)

https://huggingface.co/ngxson/dummy-llama/blob/main/llama_xs_q4.bin

I have no idea if OpenBLAS will help. You can try if you want.

phymbert committed:
server: tests: fix ci workflow

phymbert committed:
server: tests: fix ci workflow

phymbert committed:
server: tests: use ngxson llama_xs_q4.bin

ngxson commented on .github/workflows/server-test.yml:
```diff
@@ -0,0 +1,51 @@
+          mkdir build
+          cd build
+          cmake .. -DCMAKE_BUILD_TYPE=Release
+          cmake --build . --config Release -j $(nproc)
```
You can cut down the build time by adding `--target server`

phymbert commented:
> @phymbert Can you try this model instead? (pay attention to set `n_predict` or `max_tokens`, because the model never outputs EOS token)
> 
> https://huggingface.co/ngxson/dummy-llama/blob/main/llama_xs_q4.bin
> 
> I have no idea if OpenBLAS will help. You can try if you want.
> 
> @Azeirah I tried to overfit a 86M model but unfortunately it does not seem to output any of the example. But on the bright side, it outputs mostly text (not invalid bytes as I said earlier), so still usable for the test. The Q4_K_M size is only 57MB

Nice thanks [Took 0m0.481s](https://github.com/phymbert/llama.cpp/actions/runs/7965607477/job/21745462241). Note I have also reduced the KV Size.

phymbert committed:
server: tests: build only the server

phymbert committed:
server: tests: add health check and concurrent request example

phymbert commented on examples/server/tests/features/server.feature:
```diff
@@ -0,0 +1,39 @@
+      | tinyllama-2 | You are a coding assistant. | Write the fibonacci function in c++ |
+
+
+  Scenario: Health endpoint during processing with concurrent requests
```
@ngxson Here is a first example of concurrent requests, happy to have your feedback to simplify it.

ngxson replied:
Thanks. Overall it's good here, just some small suggestions regarding the naming:
- `wait for all slots processing` maybe confused (i.e. slots start or stop processing?), I'd suggest to change it to `wait N slots to become busy` where N can be 1, 2,....
- Also, for steps that check service health, can we change them to `the server health is...` ? This way we reflect the `health` keyword in step's name

phymbert replied:
Thanks @ngxson, I have tried to simplify and factorize the steps, is the new naming is clearer ?

ngxson replied:
yes it's more clear now, thanks!

ggerganov commented:
> @ggerganov @ngxson Any idea on how to improve the prompt eval time on the github runners ? Should we give a try to OpenBLAS ?

@phymbert 

Best way to improve the speed is to use as small model as possible. You can try @karpathy's tinyllamas: https://huggingface.co/karpathy/tinyllamas

Here are instructions for converting to GGUF and using in `llama.cpp`:

https://github.com/ggerganov/llama.cpp/tree/master/examples/convert-llama2c-to-ggml

For convenience, I've uploaded the smallest 260K model (~1 MB) in GGUF format here:

https://huggingface.co/ggml-org/models/blob/main/tinyllamas/stories260K.gguf

Example:

```bash
# get the model
wget https://huggingface.co/ggml-org/models/resolve/main/tinyllamas/stories260K.gguf

# run sample inference
./main -m ./stories260K.gguf -p "One day, Lily met" -n 128 -c 256

 One day, Lily met a boy named Timmy. Tim was very happy to help her mommy. He wanted to play with the ball all day. Suddenly, something unexpected happened. A little girl came over and saw a big tree. She was very sad.
Timmy wanted to play with the ball. He thought it would be fun! When he reached up, he found it st

llama_print_timings:        load time =      80.26 ms
llama_print_timings:      sample time =       1.70 ms /   128 runs   (    0.01 ms per token, 75427.22 tokens per second)
llama_print_timings: prompt eval time =       3.06 ms /     7 tokens (    0.44 ms per token,  2288.33 tokens per second)
llama_print_timings:        eval time =     134.04 ms /   127 runs   (    1.06 ms per token,   947.46 tokens per second)
llama_print_timings:       total time =     142.59 ms /   134 tokens
```

This should be ideal for CI

ggerganov commented:
Btw, one thing that would greatly improve the state of `server` in terms of debugging issues is to add detailed logs. Things like incoming requests, parameters, batch info, etc. As much information as possible should be dumped in the log file. There is some info currently saved in `llama.log`, but there should be more. 

Probably needs a separate PR to avoid this change becoming too big, but thought I would mention this in case you are interested in further helping out with maintenance

phymbert committed:
server: tests: change model to: @karpathy's tinyllamas

phymbert committed:
server: tests: refactor steps and vocabulary

phymbert committed:
server: tests: add OAI stream test, fix file end of line, fast fail behave

Azeirah commented:
For a recent PR I did, it would be nice to confirm credentialed CORS keeps working correctly too, the idea is that if you add an authorization header to your request, CORS needs to behave in a specific way.

So the tests would be something like

```behave
Scenario Outline: Credentialed CORS requests
     Given a <request> with an api token
     Then 200 OK is given back
```

```py
@step(u'a {request} with an api token')
def step_a_prompt(context):
     # add the following authorization header header to context
     # make sure headers are sent with requests
     # Authorization: Bearer asdf98afbwqo8dsjlfh
```

^^ not working, just approximate pseudocode for what the test could look like

Best for all OpenAI endpoints, but just the completion endpoint alone is most important.

phymbert committed:
server: tests: add OAI multi user scenario

netrunnereve commented:
> Best way to improve the speed is to use as small model as possible. You can try @karpathy's tinyllamas: https://huggingface.co/karpathy/tinyllamas
> 
> Here are instructions for converting to GGUF and using in `llama.cpp`:
> 
> https://github.com/ggerganov/llama.cpp/tree/master/examples/convert-llama2c-to-ggml
> 
> For convenience, I've uploaded the smallest 260K model (~1 MB) in GGUF format here:
> 
> https://huggingface.co/ggml-org/models/blob/main/tinyllamas/stories260K.gguf

This is outside the scope of this PR, but if we have such small models we should probably place them directly in the Git repository and have them run as a standard CI test. We could go through the full quantization and inference flow for each quant type and confirm that the output is correct using a fixed seed.

phymbert committed:
server: tests: slots endpoint checks

phymbert committed:
server: tests: check for infinite loops

phymbert committed:
server: tests: check for infinite loops

phymbert committed:
server: tests: add a constant seed in completion request

phymbert committed:
server: tests: add infinite loop

phymbert committed:
server: tests: add infinite loop scenario

phymbert committed:
server: tests: add infinite loop scenario

phymbert committed:
server: tests: disable issue 3969 scenario

phymbert commented on examples/server/tests/features/server.feature:
```diff
@@ -0,0 +1,100 @@
+    Then all prompts are predicted
+
+  # FIXME: #3969 infinite loop on the CI, not locally, if n_prompt * n_predict > kv_size
+  @bug
```
@ngxson @ggerganov I am facing a weird scenario where it passed locally but [not on github worker.](https://pipelinesghubeus24.actions.githubusercontent.com/jEJM6zrMhVG8NHAbd8fFWzevGGKuKhmDZUCVPMvRQzymMWYX1b/_apis/pipelines/1/runs/100/signedlogcontent/2?urlExpires=2024-02-20T22%3A20%3A33.3282517Z&urlSigningMethod=HMACV1&urlSignature=jQrLQGljIPhyP4%2FPLVhH41uU8L%2FLtMqoJ4MQQXg0vW4%3D) It looks if the total number of token to generate accross slots exceeds the kv size, the server goes to infinite loop, see #3969.

You can grep for `steps.py` in the workflow output to see that the server is busy forever while the scenario wait for idle status.

You can run locally if you replace `@bug`  by `@llama.cpp` in the feature file.

ggerganov replied:
When I run it locally, it fails earlier:

<details>

```
Feature: llama.cpp server # features/server.feature:1

  Background: Server startup  # features/server.feature:3

  @llama.cpp
  Scenario: Health                                                                               # features/server.feature:9
    Given a server listening on localhost:8080 with 2 slots, 42 as seed and llama.cpp as api key # features/steps/steps.py:10 0.000s
    Then the server is starting                                                                  # features/steps/steps.py:27 0.004s
    Then the server is healthy                                                                   # features/steps/steps.py:27
{"timestamp":1708501224,"level":"INFO","function":"log_server_request","line":2398,"message":"request","remote_addr":"127.0.0.1","remote_port":57462,"status":200,"method":"GET","path":"/health","param    Then the server is healthy                                                                   # features/steps/steps.py:27 0.011s
    When the server is healthy                                                                   # features/steps/steps.py:27
{"timestamp":1708501224,"level":"INFO","function":"log_server_request","line":2398,"message":"request","remote_addr":"127.0.0.1","remote_port":57463,"status":200,"method":"GET","path":"/health","param    When the server is healthy                                                                   # features/steps/steps.py:27 0.007s
    Then the server is ready                                                                     # features/steps/steps.py:27
{"timestamp":1708501224,"level":"INFO","function":"log_server_request","line":2398,"message":"request","remote_addr":"127.0.0.1","remote_port":57464,"status":200,"method":"GET","path":"/health","param    Then the server is ready                                                                     # features/steps/steps.py:27 0.005s
      Traceback (most recent call last):
        File "/Users/ggerganov/development/github/llama.cpp/.venv/lib/python3.11/site-packages/behave/model.py", line 1329, in run
          match.run(runner.context)
        File "/Users/ggerganov/development/github/llama.cpp/.venv/lib/python3.11/site-packages/behave/matchers.py", line 98, in run
          self.func(context, *args, **kwargs)
        File "features/steps/steps.py", line 42, in step_wait_for_the_server_to_be_started
          wait_for_health_status(context, 200, 'ok',
        File "features/steps/steps.py", line 324, in wait_for_health_status
          and (slots_idle is None or health['slots_idle'] == slots_idle)
                                     ~~~~~~^^^^^^^^^^^^^^
      KeyError: 'slots_idle'

    And all slots are idle                                                                       # None


Failing scenarios:
  features/server.feature:9  Health

0 features passed, 1 failed, 0 skipped
0 scenarios passed, 1 failed, 0 skipped, 17 untested
4 steps passed, 1 failed, 1 skipped, 0 undefined, 142 untested
Took 0m0.027s
```
</details>

phymbert replied:
Strange I have added those fields for all `200` in #5594. Have you rebased/built with the latest master ?

ggerganov replied:
Ah, I didn't notice it was picking the `server` from `build/bin`. It works now

ggerganov replied:
Running the problematic test with thread sanitizers surfaces the issue locally:

```bash
cmake -DLLAMA_SANITIZE_THREAD=ON .
```

We should add 2 more CI runs - one with thread sanitizer and one with address sanitizer enabled.

In the meantime, I will try to figure out the root cause for the failure and fix it

ggerganov replied:
Running the tests with thread sanitizer enabled we can see a few problems with the current implementation. The endpoints "/health" and "/slots" are incorrectly accessing the slot data creating a data race. Instead, such requests should be forwarded as tasks

Even if this isn't the root cause for #3969, we first need to fix this before proceeding.

It seems to me that the fix might involve a larger refactoring of `server` so it's probably better to first merge this PR (even if the CI fails) and work in a new PR to resolve the issues

phymbert replied:
Thanks for identifying the data race. As I did changes on `/health` and `/slots` endpoint in PR #5550 and #5548. I will submit a patch ASAP to use tasks. I was wrongly thinking as it's just accessing slot data in read only, no race condition will occur.

phymbert committed:
server: tests: add embeddings scenario

phymbert committed:
server: tests: add tokenize/detokenize scenario

phymbert committed:
server: tests: detokenize endpoint issue reference added

phymbert commented on examples/server/tests/features/steps/steps.py:
```diff
@@ -0,0 +1,283 @@
+    assert response.status_code == 200
+    print(response.json())
+    # FIXME the detokenize answer contains a space prefix ?
+    assert context.tokenized_text == response.json()['content'].strip()
```
@BruceMacD @ggerganov Is it normal that the `/detokenize` enpoint prefix the content with a space ? see ##3287

ggerganov replied:
Yes, it is expected - SPM tokenizer adds a whitespace prefix: https://github.com/google/sentencepiece/issues/15

phymbert commented on .github/workflows/server-test.yml:
```diff
@@ -0,0 +1,49 @@
+# Server test scenario
+name: Server Integration Tests
+
+# FIXME put only necessary triggers
```
@ggerganov Which triggers do you suggest ?

ggerganov replied:
Maybe like so:

```yml
on:
  workflow_dispatch: # allows manual triggering
  push:
    branches:
      - master
    paths: ['.github/workflows/**', '**/CMakeLists.txt', '**/Makefile', '**/*.h', '**/*.hpp', '**/*.c', '**/*.cpp']
  pull_request:
    types: [opened, synchronize, reopened]
    paths: ['**/CMakeLists.txt', '**/Makefile', '**/*.h', '**/*.hpp', '**/*.c', '**/*.cpp']
```

phymbert committed:
server: tests: simplify completion scenario

phymbert committed:
server: tests: CORS and api key checks scenario

phymbert committed:
server: tests: add a small check to be sure all started threads have generated response

phymbert commented on examples/server/tests/features/server.feature:
```diff
@@ -0,0 +1,148 @@
+      | llama-2      | You are ChatGPT.            | Say hello.                           | 64         | false            | llama.cpp | none      |
+      | codellama70b | You are a coding assistant. | Write the fibonacci function in c++. | 512        | true             | llama.cpp | none      |
+      | John-Doe     | You are an hacker.          | Write segfault code in rust.         | 0          | true             | hackme    | raised    |
+
```
@Azeirah As the server is started once for all tests, I have to include api_key in all completion requests. I did not manage to access openai request/request headers with that client version. So I cannot check the `Access-Control-Allow-Origin` is correctly set back.
Althought I think the last scenario `CORS Options` does what you were expected

Azeirah replied:
Oh the cors bug initially was caused by regardless of what the server had configured for api key. Some clients assume they're taking to real openAI server so they always send an api key, previous versions didn't handle that correctly.

For completeness sake though, the tests _should_ be able to restart the server I'd think, to verify correct operation with different parameters.

Perhaps it can be supervised by the python test script?

phymbert replied:
yea good idea, I will change the `Background` steps to actually start the server, it will be more flexible for future scenario, allowing to test different server configs.

phymbert committed:
server: tests: change CI workflow trigger

phymbert committed:
server: tests: fix the multi users infinite loop test

ggerganov committed:
server : fix initialization thread issues

phymbert committed:
server: test: ci change the GitHub workflow trigger

phymbert committed:
Merge remote-tracking branch 'origin/master' into test/server-add-ci-test

phymbert committed:
server: tests:
  * start the server at each scenario
  * split the features as each requires different server config

phymbert committed:
server: tests: ci fix model download path

phymbert committed:
server: tests: ci tests.sh exit code

ggerganov committed:
minor : fix missing new line

ggerganov committed:
ci : actually no reason to exclude GPU code from triggers

ggerganov commented on .github/workflows/server-test.yml:
```diff
@@ -0,0 +1,53 @@
+        id: server_integration_test
+        run: |
+          cd examples/server/tests
+          ./tests.sh ../../../stories260K.gguf
```
`tests.sh` no longer accepts an argument with the model file

phymbert replied:
Thanks @ggerganov, please let me few hours to ensure it's ready for review. I am currenlty fighting with multi users feature, I am facing some blocking python behavior on the client side.

ggerganov commented on examples/server/tests/features/server.feature:
```diff
@@ -0,0 +1,65 @@
+    Then  <n_predicted> tokens are predicted with content: <content>
+
+    Examples: Prompts
+      | prompt                           | n_predict | content                                                                 | n_predicted |
+      | I believe the meaning of life is | 8         | <space>going to read.                                                   | 8           |
+      | Write a joke about AI            | 64        | tion came to the park. And all his friends were very scared and did not | 32          |
+
```
Unfortunately, we cannot test for actual generated text content because this can vary from machine to machine and from model to model. Further down the road, we can think about tests similar to the approach in https://github.com/carlini/yet-another-applied-llm-benchmark. But for now we should check only deterministic stuff

phymbert replied:
I was thinking the `seed` fields does the job ? OK I will have a look.

phymbert committed:
server: fix health/slots endpoint slot state access available race condition

phymbert committed:
server: tests: switch to asyncio for concurrent tests, match result content with regex

phymbert committed:
server: tests: ci do not take a model anymore, fix trigger patch

phymbert committed:
Merge remote-tracking branch 'origin/master' into test/server-add-ci-test

phymbert committed:
server: tests: ci ensure the server is stopped before scenario, and do not quit while the server is listening

phymbert committed:
server: tests: add Multi users with total number of tokens to predict exceeds the KV Cache size

phymbert committed:
server: tests: Fix some random behavior where the wait for busy status is missing

phymbert committed:
server: tests: ci triggered on any changes on server example path

phymbert commented:
> > @Azeirah 
> Other than that, I think it's fine that the tests are in separate files. It's kinda just how behave is meant to be used, each feature is one file. Different related scenarios belong to one feature. I'm somewhat familiar with BDD myself since I use a loosely inspired variant at work, do you think BDD is unclear to some people? I could write a short readme explaining it.

@ngxson @Azeirah I finally split files per feature, as it is mandatory cucumber requirement. In order to definer server option in the the `Background` steps.

phymbert committed:
server: tests: minor fix missing param.

phymbert committed:
server: tests: assert embeddings are actually computed, make the embeddings endpoint configurable.
Add logs to investigate why the CI server test job is not starting

phymbert committed:
server: tests: minor color change

phymbert committed:
server: tests: minor fix server --alias param passed twice

phymbert committed:
server: tests: add log in server start to identify why the server does not listen on the CI

phymbert committed:
server: tests: allow to override the server port before launching tests

phymbert committed:
server: tests: ci adding container to specify server port and allow the server to listen to

phymbert committed:
server: tests: ci adding cmake as it is not present by default in ubuntu base

phymbert committed:
server: tests: ci adding python3-pip as it is not present by default in ubuntu base

phymbert committed:
server: tests: ci adding curl as it is not present by default in ubuntu base for the hf.sh script

phymbert committed:
server: tests: ci adding psmisc as it is not present by default in ubuntu base killall

phymbert committed:
server: tests: fix coloring console

phymbert committed:
server: tests: reducing n_ctx and n_predict for // prompts as it is too slow in the CI.

phymbert committed:
server: tests: reducing sleep time during scenario

phymbert committed:
server: tests: parallel fix server is started twice, add colors to help to monitor in the CI jobs

phymbert committed:
server: tests: ci : build and run tests for all matrix defines, sanitizer and type

phymbert committed:
server: tests: ci :  fix wget missing

phymbert committed:
server: tests: ci :  add git

phymbert committed:
server: tests: ci : matrix cuda

phymbert committed:
server: tests: ci : continue on error

phymbert committed:
server: tests: ci : fix step id duplicated

phymbert committed:
server: tests: ci : fix cuda install

phymbert committed:
server: test: ci fix cuda build

phymbert committed:
server: test: ci fix cublas build

phymbert committed:
server: test: ci fix matrix

phymbert committed:
server: test: ci fix clblast

phymbert committed:
server: test: ci fix openblas build

phymbert committed:
server: test: ci matrix

phymbert committed:
server: test: ci matrix, experimental on matrix avx512 entry which fail test

phymbert committed:
server: test: ci cmake remove all warning as it is done by the classical build and does not matter for testing

phymbert committed:
server: test: ci make arch not available pass the test

phymbert committed:
server: test: ax512 experimental

phymbert committed:
server: test: display server logs in case of failure

phymbert committed:
server: test: fix CUDA LD PATH

phymbert committed:
server: test: ci debug LD path

phymbert committed:
server: test: ci debug CI LD path

phymbert committed:
server: test: ci switch to nvidia based docker image for cuda

phymbert committed:
server: test: ci give up with nvidia as it requires the nvidia docker runtime

phymbert committed:
server: test: ci rename step name to Test, change matrix order for better clarity

phymbert commented:
@ggerganov I suggest to go with this first version and see how it behaves on master.

Note: sorry I messed up with github CI workflow and triggered lot of jobs on your repo actions. I am cancelling them...

@ngxson @Azeirah Happy to have your review

phymbert committed:
server: test: ci rename job name to Server

ngxson commented on examples/server/tests/features/environment.py:
```diff
@@ -0,0 +1,64 @@
+        time.sleep(0.1)
+        attempts += 1
+        if attempts > 5:
+            print(f"Server dandling exits, killing all {context.server_path} ...")
```
typo: dangling

phymbert committed:
server: tests: move all requests call to asyncio

ngxson commented on examples/server/tests/features/steps/steps.py:
```diff
@@ -0,0 +1,645 @@
+    if context.server_api_key is not None:
+        server_args.extend(['--api-key', context.server_api_key])
+    print(f"starting server with: {context.server_path}", *server_args)
+    context.server_process = subprocess.Popen(
```
Maybe quite redundant for now, but I think it's also possible to capture stdout / stderr of `server_process` maybe saved to a second file called `llama-stdout.log`. The reason is that sometimes errors are being `printf` directly instead of logging into file (for example, the `GGML_ASSERT`)

Also maybe it's worth to enable `--verbose` mode.

ngxson reviewed (approved):
Overall it's LGTM; Still I'll wait for others to leave your review. Thanks!

ggerganov reviewed (approved):
Wow! Very nice work - this would be very useful and should help to improve `server` significantly

> multi users with total number of tokens to predict exceeds the KV Cache size, fixed to be confirmed: https://github.com/ggerganov/llama.cpp/issues/3969

What was the fix?

phymbert committed:
server: tests: allow to pass argument to the test file
 add wrong_usage.feature to demonstrate user issue which will not be fixed.

phymbert committed:
server: tests: print server logs only on github action

Azeirah commented:
I will review this fully tomorrow, I'm a bit sick but I have energy when I plan it out.

phymbert committed:
server: tests: check if the server has not crashed after a scenario

phymbert committed:
server: tests: adding concurrent embedding in issue #5655
      allow to enable VERBOSE mode

phymbert commented:
> Wow! Very nice work - this would be very useful and should help to improve `server` significantly
> 
> > multi users with total number of tokens to predict exceeds the KV Cache size, fixed to be confirmed: #3969
> 
> What was the fix?

No fix was applied actually, IMHO it's a wrong usage of the server when neither `--n_predict` nor `"max_tokens"` are set. If you provide n_predict|max_tokens in the request, the server behaves well. I have updated the PR description as it was confusing.  But IMHO, a server should never infinite loop.

I have also added a  `wrong_usages.feature` file to trace and reproduce this kind user issue.

@ggerganov Regarding #5655, I have reproduced it in `issues.feature`, to run it:
`DEBUG=ON ./tests.sh --no-skipped --tags bug`

It can be investigated/fixed in another PR.

Thanks for review, I will give a last chance to concurrent streaming request with aihttp, then merging this first version.

phymbert committed:
server: tests: linter

phymbert commented:
> I will review this fully tomorrow, I'm a bit sick but I have energy when I plan it out.

@Azeirah No worries, take care, it can wait for tomorrow :+1:

phymbert committed:
server: tests: fix concurrent OAI streaming request

Azeirah commented:
> > Wow! Very nice work - this would be very useful and should help to improve `server` significantly
> > > multi users with total number of tokens to predict exceeds the KV Cache size, fixed to be confirmed: #3969
> > 
> > 
> > What was the fix?
> 
> No fix was applied actually, IMHO it's a wrong usage of the server when neither `--n_predict` nor `"max_tokens"` are set. If you provide n_predict|max_tokens in the request, the server behaves well. I have updated the PR description as it was confusing. But IMHO, a server should never infinite loop.
> 
> I have also added a `wrong_usages.feature` file to trace and reproduce this kind user issue.
> 
> @ggerganov Regarding #5655, I have reproduced it in `issues.feature`, to run it: `DEBUG=ON ./tests.sh --no-skipped --tags bug`
> 
> It can be investigated/fixed in another PR.
> 
> Thanks for review, I will give a last chance to concurrent streaming request with aihttp, then merging this first version.

In the case the server is started with undesirable parameters, we should either abort or at the very least offer a clear warning with a suggested solution. Is that the case now?

I try to focus a lot on usability for end users.

phymbert committed:
server: tests: add a note regarding inference speed.

phymbert committed:
server: tests: removing debug print

Azeirah commented on examples/server/tests/README.md:
```diff
@@ -0,0 +1,33 @@
+`pip install -r requirements.txt`
+
+### Run tests
+1. Build the server
```
Builds with makefile don't work, need to build with 

`mkdir build; cd build; cmake ../; cmake --build . --target server` from root directory. Probably works if you just create the build dir and move the makefile binary target there too, but giving an example command is nice for most people.

Azeirah commented on examples/server/tests/features/issues.feature:
```diff
@@ -0,0 +1,36 @@
+    And   a model file stories260K.gguf
+    And   a model alias tinyllama-2
+    And   42 as server seed
+    And   64 KV cache size
```
Apologies for pedantics, but KV cache size is measured in what? Some sort of domain-specific unit or is this just MB?

Azeirah reviewed (approved):
I think it's excellent! I ran it 5 times to check for flakes as well, especially since I have a capable PC and the warning in the readme, 0 flakes.

Did a quick look at all code, didn't read deep into the implementation but did look at the tests. I think it's a fantastic set-up for server tests! :)
If there's reason to add any tests later this set-up will be very easy to extend and it runs very stable so it's ready for merging if you ask me.

Only had a tiny comment about the readme again but it's very minor.

phymbert committed:
server: tests - allow print on debug

phymbert committed:
server: tests - README.md add build instruction and notice on @bug and @wrong_usage.

phymbert committed:
server: tests - add explanation about KV Cache.

phymbert committed:
server: tests - print only in case of DEBUG

phymbert committed:
server: add link to tests in the README.md

phymbert committed:
server: tests: improved README.md

phymbert committed:
github issue template: add link to the tests server framework

phymbert commented:
> Btw, one thing that would greatly improve the state of `server` in terms of debugging issues is to add detailed logs. Things like incoming requests, parameters, batch info, etc. As much information as possible should be dumped in the log file. There is some info currently saved in `llama.log`, but there should be more.
> 
> Probably needs a separate PR to avoid this change becoming too big, but thought I would mention this in case you are interested in further helping out with maintenance

On it, especially in `update_slots` as it is a nightmare to understand what's going on
